### PR TITLE
feat(web-wallet): real-MetaMask E2E infrastructure + onboarding journey

### PR DESCRIPTION
**Pull request by raul-oliveira**
## Summary

Adds a Playwright E2E layer that **installs and drives the real MetaMask Flask extension** in a real browser (headed) to validate the Web Wallet's MetaMask-Snap flows end-to-end — not a mocked `window.ethereum`.

Journeys shipped: **onboarding** (new wallet → connect + install the Hathor Snap → connected home), **import** (import the funded testnet stub → connect → switch network to testnet), and **token-lifecycle** (funded testnet: send HTR, create/send deposit + fee tokens, unregister/re-register via banner and config string), plus a **feature-example** template for adding new journeys.

> **Note.** Driving the real MV3 extension is still flaky and CI / true-headless wiring is deferred (see *Status* below); runs headed, on demand.

## What's included

**Infrastructure** (`packages/web-wallet/tests/e2e/`)
- `driver/flask.ts` — resolves/downloads a pinned MetaMask Flask build (`E2E_FLASK_VERSION` or `E2E_METAMASK_PATH`).
- `driver/selectors.ts` — version-pinned MetaMask selectors, centralized in one place.
- `driver/MetaMaskDriver.ts` — onboards/imports a wallet and drives Snap approvals via `notification.html` (MetaMask popup *windows* don't render under Playwright's persistent context). Also handles unlocking and warming the MV3 service worker.
- `driver/windows.ts` — optional side-by-side window tiling for headed runs.
- `fixtures/metamask-fixture.ts` — worker-scoped persistent context with Flask loaded; one Playwright project = one isolated wallet.
- `helpers/webWallet.ts` — dApp page object; waits for the injected provider and warms the service worker before connecting (the dApp resolves its Snaps provider once on mount).
- `helpers/journeys.ts`, `helpers/wallets.ts` — provisioning (onboard/import → connect → network) and the typed `wallets.config.json` loader.
- `playwright.e2e.config.ts` — boots the snap (`:8080`) and web-wallet (`:5173`); separate from the default Playwright config.
- `onboarding.spec.ts`, `import.spec.ts`, `token-lifecycle.spec.ts`, `feature-example.spec.ts` — the journeys (one project each) plus a template.

**Supporting changes**
- `packages/snap`: no allowlist change — origins are handled by remapping the already-allowed `https://staging.wallet.hathor.network` to the local dev server (`--host-resolver-rules`), and the E2E build uses `snap.config.e2e.ts` so `constants.ts` and the committed `snap.manifest.json` stay byte-identical.
- `packages/web-wallet`: `yarn e2e` / `yarn e2e:headed` scripts; ignore `e2e/**` from the default Playwright/Vitest/ESLint runs; `E2E_TLS` dev-server option in `webpack.config.cjs`; QA docs under `docs/`.

## How to run

```bash
cd packages/web-wallet
cp .env.e2e.example .env.e2e   # set E2E_FLASK_VERSION (e.g. 13.31.0)
yarn playwright install chromium
yarn e2e                       # headed; auto-starts the snap + web-wallet
```

Validated against **MetaMask Flask 13.31.0**.

## Status / known limitations

- ⚠️ **Flaky** (~1 in 2–3 runs): driving the real MV3 extension races MetaMask's cold service worker (connect) and Snap-dialog timing (network switch). The driver mitigates this (provider/SW warm-up, unlock handling, in-place approval waiting) but it isn't fully deterministic yet — re-running passes.
- **CI / true headless**: deferred; runs headed locally for now.
- **Funded journeys**: `import`, `token-lifecycle`, and `feature-example` use the committed `funded` testnet stub seed and run by default; clear `funded.srp` in `wallets.config.json` to skip them.
- **Permanent test data**: tokens created / txs sent persist on testnet (unique names per run; on-demand runs limit fund usage).

## Note for reviewers

This branch is stacked on the local `e2e` integration branch, so the diff against `master` currently also includes the unmerged fee-token / snap-error commits. The scope of **this** PR is `packages/web-wallet/tests/e2e/**`, `playwright.e2e.config.ts`, the `snap.config.e2e.ts` E2E build config, the `package.json` scripts, the `.gitignore`/`eslint` ignores, the `webpack.config.cjs` `E2E_TLS` option, and the `docs/`.

## Summary by CodeRabbit

* **New Features**
  * Added an opt-in Playwright E2E suite covering onboarding, importing, network switching, token lifecycle, and Snap approvals.
  * Added an E2E-only Snap serving setup with locally writable manifest/bundle handling and HTTPS-ready web-wallet runs.

* **Tests**
  * Added `e2e` and `e2e:headed` scripts plus Playwright E2E configuration with multiple wallet journeys, retries, and tuned timeouts.
  * Added richer real MetaMask/Snap driving helpers, including optional split-window UI support.

* **Documentation**
  * Added an E2E QA automation strategy guide and spec writing conventions.

* **Chores**
  * Updated E2E environment examples, E2E ignores, ESLint ignores, and Playwright test inclusion/exclusion.

### CODE REVIEW

File: packages/web-wallet/docs/qa-automation-strategy.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+# QA Automation Strategy — Web Wallet Snap Flows
+
+Companion to [`qa-snap-token-flows.md`](./qa-snap-token-flows.md). Describes how the QA
```
Comment (tuliomir) [inline packages/web-wallet/docs/qa-automation-strategy.md]: chore: This is a dead link.

File: packages/web-wallet/tests/e2e/driver/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/**
+ * MetaMask Flask selectors, centralized so the version coupling lives in ONE place.
+ *
+ * Verified by walking the live extension (Flask 13.31.0, set via E2E_FLASK_VERSION) — not
+ * assumed. If onboarding breaks after a Flask bump, re-capture these.
+ *
+ * The Snap connect/install/dialog flows happen on `notification.html` and are driven by
+ * visible role/text (see MM_TEXT) because those screens vary most between versions.
+ */
+export const MM = {
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/driver/selectors.ts]: chore: Rename this var to a more self-explanatory name. Same for the other variables in this PR with a name too simplified/compact.

```suggestion
export const MetaMask = {
```

File: packages/web-wallet/tests/e2e/e2e.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+6. **Use the real Unleash — never skip it.** Feature flags are per-network and must resolve
+   from the real proxy.
+7. **Assert only what the UI actually renders. Verify against the real component code first**
+   — never assume a value (e.g. a fee) is displayed.
+
+---
+
+## The layering model
+
+Four roles. Each has ONE job. Crossing the boundaries is what caused past confusion.
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/e2e.md]: chore: Improve wording
```suggestion
Four roles. Each has ONE job. Crossing the boundaries has caused past confusion.
```

File: packages/web-wallet/tests/e2e/e2e.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+---
+
+## Pitfalls we actually hit (don't repeat them)
+
+- ❌ Bundling the Snap approval inside a `sendToken`/`createToken` helper → can't write reject
+  tests. ✅ Keep the Snap decision in the spec.
+- ❌ Filing network-switch under "provisioning/setup" → can't express journeys that switch
+  network mid-flow. ✅ It's a spec verb.
+- ❌ Assuming the fee is shown for fee tokens → there's nothing to assert. ✅ Verify the
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/e2e.md]: suggestion: Change the "fee is not shown" concrete case. This is an example of a behavior we most probably will want to change soon. And as soon as we do, this documentation will add confusion instead of clarity.

There are other instances where this "fee is not shown" is written, and could be changed too to avoid confusion.

```suggestion
- ❌ Assuming an information is shown for a given context → it may be deliberately hidden. ✅ Verify the
```

File: packages/web-wallet/tests/e2e/e2e.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+- **Async-loaded values render a fallback first.** The Receive dialog shows
+  `No address available` until the Snap's `htr_getAddress` resolves. Don't read the value the
+  instant the dialog opens — wait for the fallback to be replaced
+  (`await expect(addr).not.toHaveText(/no address available/i)`), then read.
+- **Action buttons duplicate once tokens have balance.** There are two `Send` buttons — the
+  primary quick action (first in DOM) and a per-token Send in the asset row. `Unregister
+  token` appears twice — the trigger in the history view and the confirm button in the dialog.
+  Disambiguate with `.first()` / `.last()` (primary/quick action is first in DOM; a dialog
+  mounts on top, so its button is last) or scope to the dialog.
+- **Icon-only buttons have no accessible name.** The dialog X close and the header hamburger
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/e2e.md]: chore: This is actually an accessibility bug we should fix. Another concrete case.

We should keep this here, but just with this clarification that this is a behavior that will probably change soon, and will cause this rule/gotcha/doc obsolete.

File: packages/web-wallet/tests/e2e/import.spec.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    'Set wallets.config.json -> funded.srp to a testnet stub seed to run this journey.',
+  );
+
+  test('imports the funded wallet and connects on Mainnet', async ({ wallet }) => {
+    await wallet.expectConnected();
+    await wallet.expectNetwork('Mainnet');
+  });
+
+  test('switches the Hathor network to testnet', async ({ wallet, metamask }) => {
+    await switchNetwork(wallet, metamask, 'testnet');
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/import.spec.ts:26]: chore: There is no assertion in this test. Add one.

```suggestion
    await switchNetwork(wallet, metamask, 'testnet');
    await wallet.expectNetwork('Testnet');
```

File: packages/web-wallet/tests/e2e/onboarding.spec.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  });
+
+  test('address mode dialog closes via the X button', async ({ wallet }) => {
+    await wallet.openAddressModeDialog();
+    await wallet.closeAddressModeViaX();
+  });
+
+  test('address mode dialog closes when clicking outside', async ({ wallet }) => {
+    await wallet.openAddressModeDialog();
+    await wallet.closeAddressModeViaBackdrop();
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/onboarding.spec.ts:52]: chore: I'm also missing assertions on the last two tests. Shouldn't all of them have explicit assertion commands on the spec instead of inside helpers?

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    await pw.getByTestId(MM.onboarding.passwordNew).fill(password);
+    await pw.getByTestId(MM.onboarding.passwordConfirm).fill(password);
+    await pw.getByTestId(MM.onboarding.passwordTerms).click();
+    await pw.getByTestId(MM.onboarding.passwordSubmit).click();
+
+    const words = await this.revealAndSaveSeed();
+    await this.solveSeedQuiz(words);
+
+    // Continue triggers a "Perfect — that's right!" modal that blocks the rest.
+    await this.clickTestId(MM.onboarding.recoveryConfirm, 15_000);
```
Comment (tuliomir) [inline packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts]: chore: All those hardcoded timeouts should be centralized and configurable. This can prevent the entire suite from running on different/slower environments or cause flakiness.

File: packages/web-wallet/docs/qa-automation-strategy.md (hunk 1 above)
Comment (raul-oliveira) [reply packages/web-wallet/docs/qa-automation-strategy.md]: Good catch — `qa-snap-token-flows.md` was a planning artifact that never shipped. Removed the dead reference and pointed the "see also" at `tests/e2e/e2e.md`, which is the doc that actually exists. Fixed.

File: packages/web-wallet/tests/e2e/driver/selectors.ts (hunk 2 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/selectors.ts]: Agreed. Renamed `MM → MetaMask`, `MM_TEXT → MetaMaskText`, `SRP → Srp`, `UNLOCK → Unlock`, and updated the call sites in `MetaMaskDriver.ts`. This touches a lot of lines but it's a pure rename. Done.

File: packages/web-wallet/tests/e2e/e2e.md (hunk 3 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/e2e.md]: Applied. Thanks.

File: packages/web-wallet/tests/e2e/e2e.md (hunk 4 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/e2e.md]: Good point — tying the rule to today's "fee hidden" behavior would rot the moment we change it. Reworded to the version-independent principle ("verify the component; assert only what it renders") and swept the other "fee is not shown" mentions for the same wording. Done.

File: packages/web-wallet/tests/e2e/e2e.md (hunk 5 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/e2e.md]: Agreed, this is a real a11y bug, not just a test inconvenience. Kept the gotcha (the tests need it today) but flagged it as a workaround for a bug we should fix, so it gets removed once the buttons get accessible names. Opened #176 to track adding the `aria-label`s. Done.

File: packages/web-wallet/tests/e2e/import.spec.ts (hunk 6 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/import.spec.ts:26]: The assertion was there but buried — `switchNetwork()` ends with `expectNetwork(display)`. You're right that it's invisible from the spec. Added the explicit `expectNetwork('Testnet')` so the spec states what it verifies. (Same "assert in the spec, not inside helpers" point as your `onboarding.spec.ts` comment — I'm applying it across the suite.)

File: packages/web-wallet/tests/e2e/onboarding.spec.ts (hunk 7 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/onboarding.spec.ts:52]: You're right. The close helpers did assert internally (`toBeHidden()`), so they weren't passing silently — but the assertion wasn't visible in the spec. I split act/assert across the suite: helpers perform the action and the spec calls an explicit `expect*`. Added `expectAddressModeClosed()` for the two close tests, and while auditing the rest I found the same pattern in the copy-address test — split it into `copyFirstAddress()` + `expectClipboardHoldsFullAddress()`. Every spec now states its assertions explicitly. Done.

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts (hunk 8 above)
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts]: Agreed. The dApp-side timeouts are already env-driven on the Linux-fix branch (`SNAP_TIMEOUT_MS` in `6b30e6e`, plus a Playwright `actionTimeout` in `fb543e2`). I extended the same idea to the driver's internal deadlines: a single `TIMEOUTS` table with an `E2E_TIMEOUT_SCALE` env knob, so slower environments scale every deadline in one place. Done.

File: docs/superpowers/plans/2026-06-11-web-wallet-e2e-journeys.md [hunk 9]
```diff
@@ -0,0 +1,841 @@
...
+# wallet you want to drive. NEVER commit a real one (this file is gitignored).
+# E2E_IMPORT_SRP="word1 word2 ... word12"
+
+# --- Phases 2-3 only (token send / creation) -------------------------------
+# For send/create tests the imported wallet's Snap-derived Hathor TESTNET
+# address must be funded. Destination address for the send test (testnet):
+# E2E_DEST_ADDRESS=
+```
+with:
+```
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply docs/superpowers/plans/2026-06-11-web-wallet-e2e-journeys.md]: These planning docs aren't part of the PR — the branch was squashed and only `packages/web-wallet/docs/qa-automation-strategy.md` and `packages/web-wallet/tests/e2e/e2e.md` ship. Resolving as outdated.

File: docs/superpowers/plans/2026-06-12-web-wallet-token-lifecycle.md [hunk 10]
```diff
@@ -0,0 +1,654 @@
...
+- Playwright browser: `corepack yarn workspace @hathor/web-wallet playwright install chromium` (enableScripts:false).
+- The Playwright `webServer` block boots the Snap (`:8080`) and the dev server (`:5173`) automatically; you don't start them by hand.
+- E2E scripts live in the web-wallet workspace: `corepack yarn workspace @hathor/web-wallet e2e [args]` (alias used below: **`yarn e2e`**). `yarn e2e:headed` runs headed. `--list` lists tests without running. `E2E_DEBUG=1` dumps screens + `[notif]` buttons. `E2E_HEADED=1` forces headed; `E2E_SPLIT_WINDOWS=1` splits wallet/MetaMask into two windows.
+
+**Selector confirmation protocol (E2E against a live React UI):** the selectors below are derived from the component source, but accessible names/roles must be confirmed against the running app. Whenever a `getByRole`/`getByText`/`getByPlaceholder` fails to resolve during an integration run:
+1. Re-run that project headed with `E2E_DEBUG=1` and read the `[dump:*]` / `[notif]` output (the driver's `dumpScreen` lists testids/buttons/inputs).
+2. Or add a temporary `await page.pause()` and use the Playwright Inspector to copy the real selector.
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply docs/superpowers/plans/2026-06-12-web-wallet-token-lifecycle.md]: These planning docs aren't part of the PR — the branch was squashed and only `packages/web-wallet/docs/qa-automation-strategy.md` and `packages/web-wallet/tests/e2e/e2e.md` ship. Resolving as outdated.

File: docs/superpowers/plans/2026-06-12-web-wallet-token-lifecycle.md [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+Run (headed, only this project):
+```
+nix develop . -c bash -c 'corepack yarn workspace @hathor/web-wallet e2e:headed --project=token-lifecycle'
+```
+Expected: imports the funded wallet, connects, installs the Snap, switches to testnet → 1 passed. (If the cold-service-worker connect flake hits, re-run; it's pre-existing.) If the network dialog selectors differ, apply the selector confirmation protocol.
+
+- [ ] **Step 6: Commit (owner runs manually)**
+
+Hand to the owner:
+```
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply docs/superpowers/plans/2026-06-12-web-wallet-token-lifecycle.md]: These planning docs aren't part of the PR — the branch was squashed and only `packages/web-wallet/docs/qa-automation-strategy.md` and `packages/web-wallet/tests/e2e/e2e.md` ship. Resolving as outdated.

File: packages/web-wallet/tests/e2e/driver/flask.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  const cached = existsSync(dest) ? findManifestDir(dest) : null;
+  if (cached) return cached;
+
+  mkdirSync(dest, { recursive: true });
+  const assetUrl = await discoverFlaskAssetUrl(version);
+  const zipPath = join(CACHE_DIR, `flask-${version}.zip`);
+
+  try {
+    execSync(`curl -fL --retry 3 -o "${zipPath}" "${assetUrl}"`, { stdio: 'inherit' });
+    execSync(`unzip -oq "${zipPath}" -d "${dest}"`, { stdio: 'inherit' });
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/flask.ts]: Fair — even though `assetUrl` comes from the GitHub release API (and is regex-filtered), the array form is strictly safer and clears the linter. Switched both calls to `execFileSync`. Done.

File: packages/web-wallet/tests/e2e/driver/flask.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+ * Prefers a non-"deprecated" build when several are published.
+ */
+async function discoverFlaskAssetUrl(version: string): Promise<string> {
+  const api = `https://api.github.com/repos/MetaMask/metamask-extension/releases/tags/v${version}`;
+  const res = await fetch(api, {
+    headers: { 'User-Agent': 'hathor-web-wallet-e2e', Accept: 'application/vnd.github+json' },
+  });
+  if (!res.ok) {
+    throw new Error(`GitHub API returned ${res.status} for ${api}. Is v${version} a real Flask release?`);
+  }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/flask.ts:91]: Good call. Added a 15s `AbortController` timeout so a stalled GitHub request fails fast with a clear error instead of hanging setup. Done.

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+      if (await this.unlockIfNeeded(page)) {
+        stableUnlocked = 0;
+        await delay(1_000);
+        continue;
+      }
+      stableUnlocked += 1;
+      if (stableUnlocked >= 2) return;
+      await delay(1_200);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts:284]: Agreed — these are "ensure/finish" contracts, so a timeout should throw at the source, not leak into a later step. Added a `dumpScreen` + descriptive `throw` at the end of both loops. This pairs well with the `actionTimeout` added on the Linux-fix branch (`fb543e2`), which already makes lower-level clicks fail clearly. Done.

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+        const button = notif.getByRole('button', { name }).last();
+        if (await button.isVisible().catch(() => false) && await button.isEnabled().catch(() => false)) {
+          await button.click({ timeout: 4_000 }).catch(() => undefined);
+          break;
+        }
+      }
+      await delay(1_200);
+      if (opts.reject) return; // a single rejection resolves the request
+    }
+  }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts:584]: The idle-settle `return` is intentional ("no more dialogs"), but you're right that exhausting the hard `maxMs` cap should throw rather than report success. Added the `throw` after the loop. Note the connect path was separately hardened on the Linux-fix branch (`4f1ebb5`) with a `driveApprovalsUntil(connected)` that can't settle-and-leave. Done.

File: packages/web-wallet/tests/e2e/driver/windows.ts [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+    const { bounds: after } = await cdp.send('Browser.getWindowForTarget', {});
+    await cdp.detach().catch(() => undefined);
+    return {
+      left: after.left ?? bounds.left,
+      top: after.top ?? bounds.top,
+      width: after.width ?? bounds.width,
+      height: after.height ?? bounds.height,
+    };
+  } catch {
+    // Best-effort: tiling is a nicety; never fail a run over window placement.
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/windows.ts:155]: Right — moved `detach()` into a `finally` so the session is released even when a CDP command throws. Done.

File: packages/web-wallet/tests/e2e/fixtures/metamask-fixture.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+          '--host-resolver-rules=MAP staging.wallet.hathor.network:443 127.0.0.1:5173',
+          '--ignore-certificate-errors',
+        ],
+      });
+
+      await use(context);
+
+      await context.close();
+      rmSync(userDataDir, { recursive: true, force: true });
+    },
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/fixtures/metamask-fixture.ts:103]: Agreed. Wrapped launch/use in `try/finally` so the temp profile is always removed, even when the context fails to launch. Done. (Merges cleanly with the clipboard-permission change on the Linux-fix branch, which edits the same block.)

File: packages/web-wallet/tests/e2e/helpers/wallets.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+/** Resolve a wallet name to its entry. Throws if missing/unconfigured — guard with hasWallet. */
+export function lookupWallet(name: string): WalletEntry {
+  const entry = load()[name];
+  if (!entry?.srp) {
+    throw new Error(
+      `Wallet "${name}" is not configured in wallets.config.json (missing or empty srp). ` +
+        `Either add it, or skip the journey with test.skip(!hasWallet("${name}"), ...).`,
+    );
+  }
+  return entry;
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/helpers/wallets.ts]: Good edge case. Trimmed `srp` in both guards so a whitespace-only entry is treated as unconfigured (clear error at the source) and the returned entry carries the trimmed seed. Done.

File: packages/web-wallet/tests/e2e/helpers/webWallet.ts [hunk 19]
```diff
@@ -0,0 +1,559 @@
...
+      await this.page.getByRole('button', { name: /connect wallet/i }).click().catch(() => undefined);
+      await this.page.waitForTimeout(2_500);
+      const sustained = await this.connectingLocator().isVisible().catch(() => false);
+      if (process.env.E2E_DEBUG === '1') {
+        console.log(`[dapp] connect attempt ${attempt}: warmed=${warmed} sustained=${sustained}`);
+      }
+      if (sustained) return;
+      await this.page.reload({ waitUntil: 'domcontentloaded' }).catch(() => undefined);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/helpers/webWallet.ts:60]: Minor: `clickConnect` returning unsustained is already caught by the `expectConnected()` that immediately follows, so it never silently "passes." That said, throwing at the exact source gives a more precise message, so I added it. (The connect flow itself was also hardened on the Linux-fix branch via `driveApprovalsUntil(connected)`.) Done.

File: packages/web-wallet/tests/e2e/token-lifecycle.spec.ts [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    await wallet.openSend();
+    await wallet.fillSend({ token: 'HTR', amount: '0.01', to: addr });
+    await wallet.submitSend();
+    await metamask.confirmDialog();
+    await wallet.expectSendSuccess();
+  });
+
+  test('creates a deposit token (QADBT) and shows the 1.00 HTR deposit', async ({ wallet, metamask }) => {
+    await wallet.openCreateToken();
+    await wallet.fillCreateToken({ name: `QADBT${Date.now()}`, symbol: 'QADBT', amount: '100' });
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/token-lifecycle.spec.ts]: Agreed — testnet tokens persist, so a fixed symbol can collide with a prior run and the symbol-based selectors would target the wrong token. Switched to per-run unique symbols (`QA…`/`QB…`, ≤5 chars) captured once at the top of the `describe` and threaded through every step. Done.

File: packages/web-wallet/docs/qa-automation-strategy.md [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+## Phased delivery
+
+| Journey (Playwright project) | Scope | Status |
+|------------------------------|-------|--------|
+| `onboarding` | New wallet → connect + install Snap → connected home (Mainnet) | **✅ done — passing** |
+| `import` | Import the `funded` wallet → connect → switch to **testnet** | done — runs by default (committed `funded` stub; clear `funded.srp` to skip) |
+| `feature-example` | Template: funded wallet on testnet + stub assertion | scaffold |
+| token send (`TC-ST-001`/`TC-ST-030`) | Copy the template; funded + insufficient-balance | planned |
+| token creation (`TC-CT-001`/`TC-CT-021`) | Copy the template | planned |
+| CI integration (xvfb / true headless) | — | deferred |
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/docs/qa-automation-strategy.md]: Right, the table predated the journey landing. The journey table now lists `token-lifecycle` with its real scope (and I dropped the stale per-row status column entirely), the traceability rows point at `token-lifecycle.spec.ts`, and the architecture tree includes the two previously-missing spec files. So the "planned" labels are gone. Done.

File: packages/web-wallet/docs/qa-automation-strategy.md [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+      webWallet.ts                # page object for the dApp (verified text/roles)
+      journeys.ts                 # reusable provisioning (onboard/import -> connect -> network)
+      wallets.ts                  # typed loader over wallets.config.json
+    wallets.config.json           # committed registry of testnet stub wallets (by name)
+    onboarding.spec.ts            # journey: new wallet -> connected home
+    import.spec.ts                # journey: import funded wallet -> connect -> testnet
+    feature-example.spec.ts       # template: copy to add a feature journey
+    .cache/                       # downloaded Flask build (gitignored)
+  .env.e2e.example                # E2E config template (real .env.e2e gitignored)
+```
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/docs/qa-automation-strategy.md]: Done — added a `text` language to both directory-tree fences to clear MD040.

File: packages/web-wallet/tests/e2e/e2e.md [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+  fixtures/metamask-fixture.ts # worker-scoped mmContext / metamask / dappPage / wallet
+  helpers/
+    webWallet.ts               # page object — MetaMask-blind dApp steps
+    journeys.ts                # provisioning (import/onboard → connect → install Snap)
+    wallets.ts                 # typed loader for wallets.config.json
+  driver/
+    MetaMaskDriver.ts          # the Snap side: onboard/import/connect, confirmDialog/rejectDialog
+    windows.ts                 # optional two-window split for headed runs
+  *.spec.ts                    # one journey per file
+```
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/e2e.md]: Done — added a `text` language to both directory-tree fences to clear MD040.

File: packages/web-wallet/tests/e2e/import.spec.ts [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+test.describe.serial('import the funded wallet and switch network', () => {
+  test.skip(
+    !hasWallet('funded'),
+    'Set wallets.config.json -> funded.srp to a testnet stub seed to run this journey.',
+  );
+
+  test('imports the funded wallet and connects on Mainnet', async ({ wallet }) => {
+    await wallet.expectConnected();
+    await wallet.expectNetwork('Mainnet');
+  });
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/import.spec.ts:23]: This one I'd push back on. A freshly connected wallet already defaults to Mainnet — see `onboarding.spec.ts`, which asserts `expectNetwork('Mainnet')` with no `network` configured. `provisionImportedWallet` only switches when `network` is set, so the assertion passes on the default. Adding `network: 'mainnet'` would force a Mainnet→Mainnet switch and an extra (pointless) Snap approval. I left the config as-is and added a one-line comment noting the import test relies on the Mainnet connect-default.

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts [hunk 25]
```diff
@@ -0,0 +1,615 @@
...
+    const page = await this.openHome();
+    await page.getByTestId(MetaMask.onboarding.recoveryReveal).click();
+    await page.waitForTimeout(1200);
+    const words: string[] = [];
+    for (let i = 0; i < 12; i++) {
+      const raw = (await page.getByTestId(Srp.chip(i)).textContent().catch(() => '')) || '';
+      words.push(raw.replace(/[^a-z]/gi, ''));
+    }
+    this.seedPhrase = words.join(' ');
+    await page.getByTestId(MetaMask.onboarding.recoveryContinue).click();
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts:173]: Good catch — an empty word would corrupt the seed and surface as an opaque onboarding timeout when the quiz can't be solved. After reading the 12 chips I now `dumpScreen('recovery-phrase-missing-word')` + throw if any word is empty. Done.

File: packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts [hunk 26]
```diff
@@ -0,0 +1,615 @@
...
+    }
+  }
+
+  private async dismissGotItModal(): Promise<void> {
+    for (let i = 0; i < 12; i++) {
+      const page = await this.openHome();
+      const gotIt = page.getByRole('button', { name: MetaMaskText.gotIt }).first();
+      if (await gotIt.isVisible().catch(() => false)) {
+        await gotIt.click().catch(() => undefined);
+        return;
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply packages/web-wallet/tests/e2e/driver/MetaMaskDriver.ts]: Right — a swallowed `click()` left the "Got it" modal open and blocked the next steps. Now I only `return` when the click actually lands; otherwise it waits 300 ms and retries on the next iteration instead of exiting. Done.

**Review by tuliomir — APPROVED** [review]
I re-ran the suite from a clean cache on a slow headless Linux box and it behaves exactly as the PR honestly advertises: on-demand, headed, flaky (~1 in 2–3), not a blocking gate. The `import` journey went green; the others tripped cold-start/approval timeouts — consistent with the documented status, not a regression.

I did surface a handful of improvements while reviewing, none of which should hold up this merge. Rather than clutter the approval with inline change-requests, I've collected them — with repro steps, file references, and priorities — in a follow-up issue so a future PR can continue from a `main` where #169 has already landed: #177

**Review by pedroferreira1 — APPROVED** [review]
